arangod/Cluster/HeartbeatThread.cpp

```diff
@@ -1348,117 +1346,92 @@ bool HeartbeatThread::handlePlanChangeCoordinator(uint64_t currentPlanVersion) {
+  // instance if not yet present:
+  for (auto options : VPackObjectIterator(databases)) {
+    if (!options.value.isObject()) {
+      continue;
+    }
 
-      auto const dbName = info.getName();
+    TRI_ASSERT(options.value.get("name").isString());
+    CreateDatabaseInfo info(server(), ExecContext::current());
+    // set strict validation for database options to false.
+    // we don't want the heartbeat thread to fail here in case some
+    // invalid settings are present
+    info.strictValidation(false);
+    // when loading we allow system database names
+    auto infoResult = info.load(options.value, VPackSlice::emptyArraySlice());
+    if (infoResult.fail()) {
+      LOG_TOPIC("3fa12", ERR, Logger::HEARTBEAT)
+          << "in agency database plan for database " << options.value.toJson()
+          << ": " << infoResult.errorMessage();
+      TRI_ASSERT(false);
+    }
 
-      if (auto vocbase = databaseFeature.useDatabase(dbName);
-          vocbase == nullptr) {
-        // database does not yet exist, create it now
+    auto const dbName = info.getName();
 
-        // create a local database object...
-        [[maybe_unused]] TRI_vocbase_t* unused{};
-        Result res = databaseFeature.createDatabase(std::move(info), unused);
+    if (auto vocbase = databaseFeature.useDatabase(dbName);
+        vocbase == nullptr) {
+      // database does not yet exist, create it now
 
-        events::CreateDatabase(dbName, res, ExecContext::current());
+      // create a local database object...
+      [[maybe_unused]] TRI_vocbase_t* unused{};
+      Result res = databaseFeature.createDatabase(std::move(info), unused);
+      events::CreateDatabase(dbName, res, ExecContext::current());
 
-        if (res.fail()) {
-          LOG_TOPIC("ca877", ERR, arangodb::Logger::HEARTBEAT)
-              << "creating local database '" << dbName
-              << "' failed: " << res.errorMessage();
-        } else {
-          _hasRunOnce.store(true, std::memory_order_release);
-        }
-      } else {
-        if (vocbase->isSystem()) {
-          // workaround: _system collection already exists now on every
-          // coordinator setting _hasRunOnce lets coordinator startup continue
-          TRI_ASSERT(vocbase->id() == 1);
-          _hasRunOnce.store(true, std::memory_order_release);
-        }
+      // note: it is possible that we race with the PlanSyncer thread here,
+      // which can also create local databases upon changes in the agency
+      // Plan. if so, we don't log an error message.
+      if (res.fail() && res.isNot(TRI_ERROR_ARANGO_DUPLICATE_NAME)) {
+        LOG_TOPIC("ca877", ERR, arangodb::Logger::HEARTBEAT)
+            << "creating local database '" << dbName
+            << "' failed: " << res.errorMessage();
       }
     }
-
-  } else {
-    return false;
   }
 
-  // invalidate our local cache
-  auto& ci = server().getFeature<ClusterFeature>().clusterInfo();
-  ci.flush();
```

Comment:
Is it unnecessary now?

Reply:
Yes, the `flush()` method should only be used for testing purposes (https://github.com/arangodb/arangodb/blob/devel/arangod/Cluster/ClusterInfo.h#L359). We believe it can cause unwanted side-effects anyway, so Max and I agreed on removing it.